### PR TITLE
Enforce OAuth2 Issuer and Audience in JWT tokens #1333

### PR DESCRIPTION
**Pull request by dknoma**
Added auto realm discovery functionality. Allows for ra to enable this. This restores and enhances old functionality that was required before.

### CODE REVIEW

**Comment by dknoma** [comment]
Had to update `ControlIT` as the `oauth` config name was not up-to-date with the naming conventions of other configurations. Added a few tests to make sure that if these names go out of sync, the test will fail meaning the names aren't matching.

File: src/test/java/org/reaktivity/nukleus/oauth/internal/control/ControlIT.java [hunk 1]
```diff
@@ -46,11 +51,23 @@
...
+    {
+        assertEquals(KEYS_NAME, KEYS.name());
+    }
+
+    @Test
+    public void shouldMatchAutoDiscoverRealmsConfigName()
+    {
+        assertEquals(AUTO_DISCOVER_REALMS_NAME, AUTO_DISCOVER_REALMS.name());
+    }
+
```
Comment (jfallows) [inline src/test/java/org/reaktivity/nukleus/oauth/internal/control/ControlIT.java]: Move these test methods to `OAuthConfigurationTest` in same package as `OAuthConfiguration`, but in the test sources of course.

File: src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java [hunk 2]
```diff
@@ -79,6 +80,20 @@ private OAuthRealms(
...
+        boolean autoDiscoverRealms)
+    {
+        this.keysByKid = keysByKid;
+        if(autoDiscoverRealms)
+        {
+            for (Map.Entry<String, JsonWebKey> keyValue : this.keysByKid.entrySet())
+            {
+                realmsByName.computeIfAbsent(keyValue.getKey(), OAuthRealm::new);
+            }
+        }
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java:91]: Add a method called `supplyRealm` as follows:
```
OAuthRealm supplyRealm(
    String realmName)
{
    return realmsByName.computeIfAbsent(realmName, OAuthRealm::new);
}
```
...then use it in `OAuthRealms` constructor.
```
if (autoDiscoverRealms)
{
    keysById.keySet().forEach(this::supplyRealm);
}
```

File: src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java [hunk 3]
```diff
@@ -151,7 +166,6 @@ public JsonWebKey lookupKey(
         Path keyFile)
     {
         Map<String, JsonWebKey> keysByKid = Collections.emptyMap();
-
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/oauth/internal/OAuthRealms.java]: Unintentional whitespace diff?

File: pom.xml [hunk 4]
```diff
@@ -52,7 +52,7 @@
 
     <nukleus.plugin.version>0.33</nukleus.plugin.version>
 
-    <nukleus.oauth.spec.version>0.24</nukleus.oauth.spec.version>
+    <nukleus.oauth.spec.version>develop-SNAPSHOT</nukleus.oauth.spec.version>
```
Comment (jfallows) [inline pom.xml]: Are there any corresponding `nukleus-oauth.spec` changes?
If so, please include a link to the PR.

**Review by jfallows — CHANGES_REQUESTED** [review]

**Review by jfallows — APPROVED** [review]

**Pull request merged**